tests/FileStorageTest.php

```diff
@@ -1,38 +1,299 @@
+    {
+        if (!ENABLE_DEV_TESTS) {
+            $this->markTestSkipped('just for development');
+        }
+
+        $storage = $this->client->createS3FileStorage(self::DEFAULT_S3_OPTIONS);
+
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage(sprintf('File storage with id "%d" and provider "azure" was not found.', $storage['id']));
+        $this->client->setAbsFileStorageAsDefault($storage['id']);
+    }
+
+    public function testCrossProviderStorageCredentialsRotateS3Abs()
+    {
+        if (!ENABLE_DEV_TESTS) {
+            $this->markTestSkipped('just for development');
+        }
+
+        $storage = $this->client->createS3FileStorage(self::DEFAULT_S3_OPTIONS);
+
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage(sprintf('File storage with id "%d" and provider "azure" was not found.', $storage['id']));
+        $this->client->rotateAbsFileStorageCredentials($storage['id'], self::ROTATE_ABS_OPTIONS);
+    }
+
+    public function testCrossProviderStorageCredentialsRotateAbsS3()
+    {
+        if (!ENABLE_DEV_TESTS) {
+            $this->markTestSkipped('just for development');
+        }
+
+        $storage = $this->client->createAbsFileStorage(self::DEFAULT_ABS_OPTIONS);
+
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage(sprintf('File storage with id "%d" and provider "aws" was not found.', $storage['id']));
+        $this->client->rotateS3FileStorageCredentials($storage['id'], self::ROTATE_S3_OPTIONS);
+    }
+
+    public function testCreateS3StorageWithoutRequiredParam()
+    {
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage(json_encode([
+            "[awsKey] : This field is missing.",
+            "[awsSecret] : This field is missing.",
+            "[filesBucket] : This field is missing.",
+            "[owner] : This field is missing.",
+            "[region] : This field is missing.",
+            ]
+        ));
+        $this->client->createS3FileStorage([]);
+    }
+
+    public function testCreateAbsStorageWithoutRequiredParam()
+    {
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage(json_encode([
+            "[accountName] : This field is missing.",
+            "[accountKey] : This field is missing.",
+            "[containerName] : This field is missing.",
+            "[owner] : This field is missing.",
+        ]));
+        $this->client->createAbsFileStorage([]);
+    }
+
+    public function testRotateS3CredentialsWithoutRequiredParams()
+    {
+        if (!ENABLE_DEV_TESTS) {
+            $this->markTestSkipped('just for development');
+        }
+
+        $storage = $this->client->createS3FileStorage(self::DEFAULT_S3_OPTIONS);
+
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage(json_encode([
```

Comment:
Chybový hlášky takhle ne. Nikde se takhle nepoužívají, je to divnej a řekl bych i špatnej formát. Nepoužíváme hranatý závorky (nebo jsem si toho nevšiml) pro encapsulaci názvů a okolo dvojtečky určitě nejsou mezery na obou stranách. Nemáme na to žádný guidelines (nebylo to jako téma na offsite?), ale já bych použil něco jako

> Missing fields "awsKey", "awsSecret". 

> Required fields are "awsKey", "awsSecret". 

Zároveň si nejsem jistej, jestli se props v jsonu dá říkat fields, JS objekty mají `property`, wiki u JSONu říká `key`.

Současná error response Manage API vypadá takhle

```json
{
  "error": "Invalid acccess token",
  "exceptionId": "connection-7704a8fcf4c51e5d22e247059b9fab55",
  "code": "storage.tokenInvalid"
}
```
Tohle by nemělo vypadat jinak, tj. `error`, `exceptionId` i `code` jsou stringy, nic jinýho, takže exception message by neměl bejt serializovanej JSON.

```json
{
    "error": "[\"[awsKey] : This field is missing.\",\"[awsSecret] : This field is missing.\"]",
    "exceptionId": "connection-7704a8fcf4c51e5d22e247059b9fab55",
    "code": "storage.invalidRequest"
}
```
je teda zvěrstvo na tolika úrovních, že to ani nebudu rozepisovat, takže doufám, že jsem se v tom reverse engineeringu spletl :-)

Reply:
Jo to jsem chtěl právě vyřešit já, protože si myslím, že chybová hláška ve stringu "Missing required params a,b,c" je takový oldies style trochu. Protože nejdřív validuješ jestli jsou všechny required... Pak A má být číslo, ale ty si poslal string...  To ti projde a pak validuješ jestli hodnota `A < 5 && A >1`, ale A je 10, takže vyhodíš další chybovou hlášku.

Ten způsob jaký jsem to udělal, ti tohle všechno napíše najednou. "Request obsahuje chybné hodnoty a jsou to tyhle":
 - `B` - chybí úplně
 - `A` - má být číslo
 - `C` - má být alespoň 10 znaků dlouhé.

Reply:
Nehledě na to, že se to pak krásně řeší v kódu: https://github.com/keboola/connection/pull/2041/files#diff-1c8588ed4ad55f6e21ec6476642fc02fR123

Reply:
Tohle je špatnej argument pro to serializovat pole do stringu v JSONu :-) Jak jsi někde psal vejš, používáme to my, takže UX úplně řešit nemusíme. Mnohem víc uživatele API pálí, že třeba SAPI není v JSONu, to nikdo nedá na první dobrou a tipnu si, že ani nikdo z nás v curlu nedá nějakej komplikovanější api call z hlavy správně.

Reply:
No jasný. Protože mi příjde lepší, když ti to vysype všechny hlášky najednou. Klidně to hodím do stringu místo JSON.

Spíš mi jde to o jestli to má být:
 - Chybí ti "A,B,C". Doplníš "A, B ,C". "A" je max 10. Změníš "A" na 10.
Nebo můj přístup:
 - Chybí ti "A", Chybí ti "B", Chybí ti "C", "C" má být od 0 do 10.

To jsem chtěl právě vyřešit na offsite :confused:

Reply:
Udělej error: "invalid request" a violations: pole těch věcí `a: ['fail 1', 'fail 2], b: ['fail']`. Hlavně je špatně, že když uděláš json_decode na té odpovědi, tak dosaneš `error: json_encode([pole])` a ne `error: [pole]`

Reply:
```
try {
        $this->client->createAbsFileStorage([]);
            } catch (ClientException $e) {
            $params = $e->getContextParams();

            $this->assertSame('invalid request', $params['error']);
            $this->assertSame([...], $params['violations']);
        }
```

Reply:
`violations` je možná moc konkrétní, nebylo by něco obecnějšího, co můžeš použít, když to není violation, ale třeba pole chyb ve z transformací? `error (string)` a `errors (array)`? tohle bych taky možná odložil a řekl si, jak to budeme dělat. na offsite jsme to prošvihli, tak mi jako nejlepší přijde, když uděláš RFC a ten s nějakou omezenou skupinou vypilujete a zbytku naservírujete. vyjet si chybový hlášky, co teď přes api vracíme, je piece of cake, takže bys mohl mít přehled, co se všechno reportuje za chyby a jak moc je to blbě.

Reply:
bylo by super, kdyby `error (string)` zůstal zachovanej, protože s ním třeba UI umí pracovat.

Reply:
jo souhlas. To bylo spíš moje nepochopení v tom, že chybový hlášky mají být JSON, tak jsem to takhle blbě zabalil a napsal, že mi to příjde divný :smiley:

Reply:
Ok změnil jsem to. je to string a vypadá to nějak takhle:
```
accountName": This field is missing. "accountKey": This field is missing. "containerName": This field is missing. "owner": This field is missing.
```

Reply:
Umí to teď jinou validaci než `This field is missing`?

Reply:
Ano https://symfony.com/doc/current/validation.html#constraints

Reply:
Umí to úplně všechno z fleku. Akorát je to dělaný na formáluře, takže se ty errory vrací strukturovaně tak, aby si je pak mohl zpracovat na formulářový prvky a né jako API response. Ale to je jenom o nějaký mezi vrstvě, která to bude správně parsovat.

Reply:
Jenom potřebuju vědět tu textaci ...
má to být "Chyba A je ve fieldech 1,2,3"
nebo "Ve fieldu 1 je chyba A,B,C ..." atd.

Každopádně bych rád upustil od validace typu `array_diff`, `if value > 5` ... Tady je to už dávno všechno vyřešený objektově a přehledně.